admin/runner.py

```diff
@@ -3,20 +3,77 @@
 Tools for running commands.
 """
 import sys
-from pipes import quote
-from subprocess import check_call, CalledProcessError
+import os
 
-def run(command, **kwargs):
+from characteristic import attributes
+
+from twisted.internet.error import ConnectionDone
+from twisted.internet.endpoints import ProcessEndpoint, connectProtocol
+from twisted.internet.defer import Deferred
+
+from twisted.protocols.basic import LineOnlyReceiver
+
+
+@attributes([
+    "deferred",
+    "output",
+])
+class CommandProtocol(LineOnlyReceiver, object):
     """
-    Echo and run a command..
+    Protocol that logs the lines of a remote command.
 
-    :param list command: Command to run.
-    :param kwargs: Extra args to pass to ``subprocess.call``.
+    :ivar Deferred deferred: Deferred to fire when the command finishes
+        If the command finished succesfuly, will fire with ``None``.
+        Otherwise, errbacks with the reason.
+    :ivar file-like output: For logging.
     """
-    sys.stdout.write("Running %s\n" % (b' '.join(map(quote, command))))
-    try:
-        check_call(command, **kwargs)
-    except CalledProcessError as e:
-        sys.stderr.write('Failed %d: %s\n'
-                         % (e.returncode, ' '.join(map(quote, command))))
-        raise SystemExit(e.returncode)
+    delimiter = b'\n'
+
+    def connectionMade(self):
+        self.transport.disconnecting = False
+
+    def connectionLost(self, reason):
+        if reason.check(ConnectionDone):
+            self.deferred.callback(None)
+        else:
+            self.deferred.errback(reason)
+
+    def lineReceived(self, line):
+        self.output.write(line + "\n")
+
+
+def run(reactor, command, **kwargs):
+    """
+    Run a process and kill it if the reactor stops.
+
+    :param reactor: Reactor to use.
+    :param list command: The command to run.
+
+    :return Deferred: Deferred that fires when the process is ended.
+    """
+    if 'env' not in kwargs:
+        kwargs['env'] = os.environ
+    endpoint = ProcessEndpoint(reactor, command[0], command, **kwargs)
+    protocol_done = Deferred()
+    protocol = CommandProtocol(deferred=protocol_done, output=sys.stdout)
+
+    connected = connectProtocol(endpoint, protocol)
+
+    def unregister_killer(result, trigger_id):
+        try:
+            reactor.removeSystemEventTrigger(trigger_id)
+        except:
+            # If we can't remove the trigger, presumably
```

Comment:
If possible, specify the exceptions which fit this presumption and then if anything else is thrown we can deal with the unexpected case.